### DISCUSSION TITLE
Allow pattern matching for object color attributes (or use the underlying index color)

### DISCUSSION THREAD

**Discussion by protist** [neomutt/neomutt#4498] [Development]
##  What would you like to be able to do?

I have my index coloured in an informative way. For example, I highlight messages with attachments.

```
color index color91 color229 "~X 1-"
```
However, the active email is highlighted in a particular colour, which overrides this attachment colour. Hence, if I select an email, I cannot easily tell if it has an attachment or not. I can only see the attachment status for unselected emails. (I know I can use `set arrow_cursor`, but this makes it hard to follow across on long lines.)

Ideally, the indicator colour should not hide the attributes of the selected message.

## What needs to change in NeoMutt?

One way to fix this would be to allow pattern matching for _objects_, specifically the indicator. Then I could do something like this:
```
color indicator green red "~X 1-"
```

Another option would be for the _attribute_ of the object to modify the "proper" underlying colour.  i.e. I would do the following:
```
color indicator bold default default
```
Here `default` (or another variable) would represent the underlying colour. Hence, selected emails with attachments would retain their colouring, but with a bold font. At the moment, `default` appears to revert back to the general (non-attachment) formatting for the index, albeit bold.

## Have you tried any alternative methods?

I have not tried macros, hooks, nor scripts. I'm not sure how that could work in this case, but any suggestions are welcome!

## Additional Info

I'm using the latest neomutt-git on Arch Linux: 20180716.r2004.g940a297f1-1.

**Comment by flatcap** [comment]
Unfortunately, this is a limitation of how (neo)mutt uses curses for colour.

Each colour is defined as a fg/bg pair, then we use this palette of colours on screen.
If we wanted to combine the two colours: `red, default` and `default, green`
we'd actually have to create a third colour `red, green`.
We do this in the index to make sure that the tree characters adopt the cursorline colour.

I think the palette allows 32767 colour pairs, so we _could_ make it work, but it'd be a lot of effort.

---

One of my attempts to work around this problem was the `lolcat` branch in the [neomutt-old](https://github.com/neomutt/neomutt-old) repo.
I sidestep curses completely and use ANSI escape sequences to allow us any combination of colour and attributes.

Screencast: https://asciinema.org/a/160886

**Comment by protist** [comment]
Thank you for the reply. I'm guessing that you are referring to the difficulties in modifying the "underlying" index colour? I don't entirely understand, though. Are you saying that the problem is mapping the third colour from the first two?

Might the first option be simpler to code? i.e. just adding pattern matching to objects (e.g. the indicator)?

Thanks again.

**Comment by flatcap** [comment]
> just adding pattern matching to objects (e.g. the indicator)?

For each place where we combine colour pairs, we'd need a new colour pair.

For every colour pair that you used in the index, we'd need an 'indicator' version of that colour too.  That's just the simple case.

Then consider 10 subject colour rules that optionally set either the foreground or background.
Suddenly the number of colour pairs we need, explodes.

**Comment by protist** [comment]
Ah okay, got it. I don't really understand enough about how the code works, but is that a problem though? I mean, couldn't you just programmatically map the "indicator" version? e.g. create a new colour on-the-fly that is the same the same hue, but bold, or whatever? (Also, I like that we both spell "colour" correctly 😁)

And alternatively:
> Might the first option be simpler to code? i.e. just adding pattern matching to objects (e.g. the indicator)?

**Comment by flatcap** [comment]
> create a new colour on-the-fly

We _could_, but I'd prefer _not_ to go down that route.
Mutt's code contains too many short-term solutions.

Many modern 'curses' apps don't actually use curses any more and choose to do the drawing themselves.
My lolcat branch tested a half-step towards that goal.

**Comment by protist** [comment]
Ah okay that makes sense then.

**Comment by fullstopslash** [comment]
What about doing something similar to the `index_flags` option, ie. `indicator_flags` ?

**Comment by protist** [comment]
@fullstopslash I'm not sure if you are talking to me or if it's a general comment for this issue, but I can confirm that the "active email" highlight also overwrites the `index_flags` highlighting. Ideally neomutt would also allow fine control of this, too, allowing an option to reveal the "underlying" colour, or to define a new custom colour.
